ocpi-toolkit-2.2.1/src/main/kotlin/com/izivia/ocpi/toolkit/modules/cdr/repositories/CdrsEmspRepository.kt

```diff
@@ -0,0 +1,43 @@
+package com.izivia.ocpi.toolkit.modules.cdr.repositories
+
+import com.izivia.ocpi.toolkit.common.URL
+import com.izivia.ocpi.toolkit.modules.cdr.domain.Cdr
+
+/**
+ * Typically implemented by market roles like: eMSP.
+ *
+ * The CDRs endpoint can be used to create and retrieve CDRs.
+ *
+ * - GET: Retrieve an existing CDR.
+ * - POST: Send a new CDR.
+ * - PUT: n/a (CDRs cannot be replaced)
+ * - PATCH: n/a (CDRs cannot be updated)
+ * - DELETE: n/a (CDRs cannot be removed)
+ */
+interface CdrsEmspRepository {
+    /**
+     * GET Method
+     *
+     * Fetch CDRs from the receivers' system.
+     *
+     * Endpoint structure definition:
+     *
+     * No structure defined. This is open to the eMSP to define, the URL is provided to the CPO by the eMSP in
+     * the result of the POST request. Therefore, OCPI does not define variables.
+     *
+     * @return Cdr Requested Cdr object.
+     */
+    suspend fun getCdr(url: URL): Cdr?
```

Comment:
I do not think this should be in `CdrsEmspRepository` because it would force users of the lib to implement this method, which does not make sense as receiver

You can remove it, and remove `override` keyword in `CdrsCpoClient`

Reply:
I HATE the fact that it's not clearly defined by the OCPI standard

Reply:
The documentation defines clearly this endpoint. Which is not defined is if it's mandatory or not . . .

Reply:
The only sentence about this is:

> If the CPO, for any reason, wants to view a CDR it has posted to an eMSP’s system, the CPO can retrieve the CDR by performing a GET request on the eMSP’s CDRs endpoint at the URL returned in the response to the POST.

Reply:
The issue here is that it's free to the user to define the url to access a CDR.

It means that we can not define the endpoint in `CdrsEmspServer.kt` in the lib (https://github.com/IZIVIA/ocpi-toolkit/pull/109#discussion_r1644457103)

The other remark is about removing `suspend fun getCdr(url: URL): Cdr?` from the interface. It does not make sense for the user to implement `getCdr` that takes an url in parameter.

`getCdr` should take an id in parameter or a date, or any identifier, but not a URL

---

Another solution would be to add a generic in `CdrsEmspRepository<T>`, so that the client will implement `CdrsEmspRepository<URL>`, and the user will implement `CdrsEmspRepository<String>` (or something else). It will still be up to the user to implement its handler in `CdrsEmspServer`

I don't really like this idea

---

And yet another solution would be to leave the responsibility to the lib to define the endpoint for the user of the lib. I think it could be nice.

For that, we will need a generic in `CdrsEmspRepository<T>`. Then:
- `CdrsCpoClient.kt` will implement `CdrsEmspRepository<URL>` (the implementation stays the same)
- `CdrsEmspService.kt` will implement `CdrsEmspRepository<String>` (the implementation stays the same)
- `CdrsEmspServer.kt` will have to be updated to assume that the last pathParam is the CDR id

Reply:
In the ChargingProfiles module, the callback URLs are defined by the lib (`ChargingProfilesScspClient.kt` & `ChargingProfilesScspServer.kt`)

Reply:
In our case, the callback is passed to the client through the param of `getCdr(url)`.

Concerning the server, we can either leave the responsibility to the user of the lib to return `baseUrl/2.2.1/cdr/{cdrId}` in its `CdrsEmspInterface` implementation. Or we can code it in the lib in `CdrsEmspServer` (the POST cdr handler)

Reply:
Coding it in the lib can be helpful. If one user need another implementation, they can implement their own.

Reply:
Finally it is not technically possible to code it in the lib. In that case we would override what's returned by postCdr by the user, and it's not really explicit for the user of the lib (and they may think it's a bug)

The best alternative is to let the user specify the URL